### PR TITLE
Auto unroll generators produced by jinja filters

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Auto unroll generators produced by jinja filters

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
```
lib/ansible/template/__init__.py
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/template/__init__.py [hunk 1]
```diff
@@ -231,6 +232,30 @@ def recursive_check_defined(item):
...
+    and auto unrolls a generator, so that users are not required to
+    explicitly use ``|list`` to unroll.
+    """
+    def wrapper(*args, **kwargs):
+        ret = func(*args, **kwargs)
+        if USE_JINJA2_NATIVE and isinstance(ret, types.GeneratorType):
+            return list(ret)
+        return ret
+
+    for attr in ('__module__', '__name__', '__qualname__', '__doc__', '__annotations__'):
```
Comment (sivel) [inline lib/ansible/template/__init__.py:261]: Need to add a comment about why I'm not just using `functools.wraps` or `functools.update_wrapper`. This is a copy from py3.7, that doesn't traceback.

File: lib/ansible/template/__init__.py [hunk 2]
```diff
@@ -726,6 +767,16 @@ def _finalize(self, thing):
...
             return thing
+
+        if _is_rolled(thing):
+            # Auto unroll a generator, so that users are not required to
+            # explicitly use ``|list`` to unroll
+            # This only affects the scenario where the final result of templating
+            # is a generator, and not where a filter creates a generator in the middle
+            # of a template. See ``_unroll_iterator`` for the other case. This is probably
+            # unncessary
+            return list(thing)
```
Comment (mkrizek) [inline lib/ansible/template/__init__.py:872]: I was thinking of doing this for native jinja too, either here in `_finalize` or in `ansible_native_concat`, to solve https://github.com/ansible/ansible/issues/59019.

Comment (sivel) [reply lib/ansible/template/__init__.py:872]: I could probably move this above the jinja2 native check in that case then.  I just figured I wouldn't touch native, but I can see a benefit.

File: lib/ansible/template/__init__.py [hunk 3]
```diff
@@ -43,9 +43,10 @@
 
 from ansible import constants as C
 from ansible.errors import AnsibleError, AnsibleFilterError, AnsibleUndefinedVariable, AnsibleAssertionError
-from ansible.module_utils.six import iteritems, string_types, text_type
+from ansible.module_utils.six import PY3, iteritems, string_types, text_type
```
Comment (s-hertel) [inline lib/ansible/template/__init__.py]:
```suggestion
from ansible.module_utils.six import iteritems, string_types, text_type
```

**Review by s-hertel — APPROVED** [review]
LGTM. This could probably use a changelog/porting guide entry.

File: lib/ansible/template/__init__.py (hunk 3 above)
Comment (sivel) [reply lib/ansible/template/__init__.py]: Hm, not sure if that was due to a mad rebase, or whether I used it in the past.  I'll get it removed.

File: lib/ansible/template/__init__.py (hunk 2 above)
Comment (sivel) [reply lib/ansible/template/__init__.py:872]: Done.

**Review by mkrizek — APPROVED** [review]
Fixes https://github.com/ansible/ansible/issues/59019 too :+1:

**Comment by bcoca** [comment]
code lgtm, but i would ensure ci_complete run before merge

**Review by bcoca — APPROVED** [review]

**Review by samdoran — APPROVED** [review]

File: lib/ansible/template/__init__.py [hunk 4]
```diff
@@ -928,6 +979,8 @@ def do_template(self, data, preserve_trailing_newlines=True, escape_backslashes=
 
             # Adds Ansible custom filters and tests
             myenv.filters.update(self._get_filters())
+            for k in myenv.filters:
+                 myenv.filters[k] = _unroll_iterator(myenv.filters[k])
```
Comment (samdoran) [inline lib/ansible/template/__init__.py]:
```suggestion
                myenv.filters[k] = _unroll_iterator(myenv.filters[k])
```

Comment (samdoran) [reply lib/ansible/template/__init__.py]: That's the test failure.

**Pull request merged**